### PR TITLE
Array constructors now accept a size and filler parameters.

### LINKED ISSUE #2181: Ability to init an array of specified size, pre-filled with a specified init value
### Describe the project you are working on
I'm working on a large astronomy simulation ([I, Voyager](https://github.com/ivoyager)) and an associated space 4X game.

### Describe the problem or limitation you are having in your project
I find myself working with very many arrays of fixed size that need to be initialized with the value 0.0, or sometimes 0. It's only a few lines of code to do this (or really just a call to a custom static function). But if others do this even a fraction as much as I do, then it perhaps should be a core capability.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
I would like to be able to init an array of specified size, pre-filled with a specified init value - using a core function. This is purely a convenience proposal.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
To init an array of size 10, pre-filled with value 0.0, this is what I'd like to work:

`var array := Array(10, 0.0)`

Edit: Based on discussion below, this proposal applies also to PackedArrays. I'm also very supportive of the `Array.fill(0.0)` method in addition to the constructor.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
I suspect this will be used often. If I'm wrong about that, then yes, it can be worked around with a few lines of code. 

### Is there a reason why this should be core and not an add-on in the asset library?
I don't think it can be done outside of a core change.

### PR DESCRIPTION
**Pull request by DarknessCatt**
Pull Request for godotengine/godot-proposals#2181.

These changes improve the Array constructor, adding a size and filler parameters.

Examples:
`var arr := Array(3) #[null, null, null]`
`var arr := Array(5, 3) #[3,3,3,3,3]`

Details:
Most variant types are supported as fillers, with a few expections: RID, StringName, NodePath, Callable and Signal are not supported, since most are unique so I considered it a bit weird to fill an entire array with the "same object".  I didn't manage to allow Objects (a lot of compiler problems), so they are not support (and, since they don't seem to have a "duplicate", they would just be references to the same object).

When arrays, packedArrays and dictionaries are fillers, they are deep copied.

I would also like to thank [anberino](https://github.com/anberino). He ran all tests for me (since my gpu doesn't support vulkan) and this PR wouldnt be possible without his help.

*Bugsquad edit: This closes godotengine/godot-proposals#2181.*

### CODE REVIEW

**Comment by pycbouh** [comment]
You have formatting issues which trigger an error in CI. Please consider installing [pre-commit hooks](https://github.com/godotengine/godot/tree/master/misc/hooks) to tests these things locally before pushing.

File: core/variant/variant_construct.cpp [hunk 1]
```diff
@@ -717,6 +717,35 @@ void Variant::_register_variant_constructors() {
 
 	add_constructor<VariantConstructNoArgs<Array>>(sarray());
 	add_constructor<VariantConstructor<Array, Array>>(sarray("from"));
+	add_constructor<VariantConstructor<Array, int>>(sarray("size"));
+	add_constructor<VariantConstructor<Array, int, bool>>(sarray("size", "fill"));
+	add_constructor<VariantConstructor<Array, int, int>>(sarray("size", "fill"));
+	add_constructor<VariantConstructor<Array, int, float>>(sarray("size", "fill"));
```
[the remark this answers is not in the document]
Comment (DarknessCatt) [reply core/variant/variant_construct.cpp:723]: It was the first thing I tried to be honest. From what I remember, the problem is that there's no `GetVariantType<Variant>` (it makes sense), so it doesn't compile.

I will change it to int64_t and doubles, as well as adding the documentation.

File: core/variant/array.cpp [hunk 2]
```diff
@@ -556,6 +556,26 @@ Array::Array() {
...
+	_p = memnew(ArrayPrivate);
+	_p->refcount.init();
+	resize(size);
+	if (fill.get_type() < Variant::DICTIONARY) {
+		for (int i = 0; i < size; i++)
+			set(i, fill);
+	} else {
+		for (int i = 0; i < size - 1; i++)
+			set(i, fill.duplicate(true));
+		set(size - 1, fill);
```
Comment (kleonc) [inline core/variant/array.cpp:575]: I think duplicating `fill` for all of the entries besides the last one might result in an unexpected behaviour. I'd say be consistent and duplicate `fill` for all entries.
Also it should be documented when `fill` is being duplicated.

File: core/variant/array.cpp [hunk 3]
```diff
@@ -556,6 +556,26 @@ Array::Array() {
...
+Array::Array(int size, const Variant &fill) {
+	_p = memnew(ArrayPrivate);
+	_p->refcount.init();
+	resize(size);
+	if (fill.get_type() < Variant::DICTIONARY) {
+		for (int i = 0; i < size; i++)
+			set(i, fill);
+	} else {
+		for (int i = 0; i < size - 1; i++)
+			set(i, fill.duplicate(true));
```
Comment (aaronfranke) [inline core/variant/array.cpp:574]: Braces are required.

```suggestion
		for (int i = 0; i < size; i++) {
			set(i, fill);
		}
	} else {
		for (int i = 0; i < size - 1; i++) {
			set(i, fill.duplicate(true));
		}
```

**Comment by DarknessCatt** [comment]
Everything commented should be fixed by now.

The new constructors have been added to the Array class documentation. In the case of values that are duplicated (Dictionary, Array and PackedArray), there's a note saying that the values are deep-copied.

File: doc/classes/Array.xml [hunk 4]
```diff
@@ -52,6 +52,334 @@
 				Constructs an empty [Array].
 			</description>
 		</method>
+		<method name="Array" qualifiers="constructor">
+			<return type="Array">
+			</return>
+			<argument index="0" name="size" type="int">
+			</argument>
+			<description>
+				Constructs an empty [Array] with an initial size.
```
Comment (akien-mga) [inline doc/classes/Array.xml:61]:
```suggestion
				Constructs an [Array] with an initial size and uninitialized values.
```

File: core/variant/array.h [hunk 5]
```diff
@@ -113,6 +113,8 @@ class Array {
 	void set_typed(uint32_t p_type, const StringName &p_class_name, const Variant &p_script);
 	Array(const Array &p_from);
 	Array();
+	Array(int size);
+	Array(int size, const Variant &fill);
```
Comment (akien-mga) [inline core/variant/array.h:117]:
```suggestion
	Array(int p_size);
	Array(int p_size, const Variant &p_fill);
```
Same in the `.cpp`.

**Comment by reduz** [comment]
As discussed with @akien-mga, this is not something you will need often, so there is no need to bloat constructors or make them confusing for the sake of it.

I would recommend maybe an Array.fill(value) function instead.

**Comment by DarknessCatt** [comment]
My original ideia was simply adding a `add_constructor<VariantConstructor<Array, int, Variant>>(sarray("size", "fill"));`, but I couldn't bind the Variant type (perhaps due to my lack of knowledge of C++). That's why I ended up making all those constructors.

What's the next step, then? Should I just close this PR and open another one with the `Array.fill(value)`? Following the original issue's problem, should I keep the size constructor (Avoiding having an unecessary call  of `Array.resize(size)` before the `Array.fill(value)`)?

**Comment by akien-mga** [comment]
The sized constructor seems to be exactly the same as doing:
```
var arr = []
arr.resize(size)
```
so I'm not sure how useful it is.

It can be considered but that's purely syntactic sugar, which might be misunderstand for creating an array with a single element with value `size` (since e.g. `Array()` is equivalent to `[]`, one could think that `Array(10)` is equivalent to `[10]`).

So I'd suggest making a PR to implement `fill()` (which might also be meaningful for packed arrays with their respective type?), and for `Array(p_size)` this likely warrants more discussion.

**Comment by DarknessCatt** [comment]
The original issue was that, while you could work around this constructor by doing:

```
var arr = Array()
arr.resize(5)
for i in range(arr.size()):
    arr[i] = value
```

It is slower.

However, I understand your points and agree the loads of constructors are quite ugly (specially bloating the documentation), so I will close this PR and open another one with the "fill" method.

**Pull request closed without merging**